### PR TITLE
Support nested categories

### PR DESCRIPTION
**Pull request by sungshik**
This PR adds support for nested categories to the converter. Concretely:

  - There is a new module `lang::rascal::grammar::analyze::Categories` that has some functions to analyze which categories might be active for each production
  - There is a new test module to test combinations of nested categories

Also:

  - One function was renamed (new name: `prodsOf`) to make its name more consistent with another new function. It's used throughout the project, though, so there are a few extra files changed because of this.
  - I enabled "Trim Trailing Whitespace", so maybe better to turn off whitespace in the diffs.

### CODE REVIEW

File: rascal-textmate-core/src/main/rascal/lang/rascal/grammar/Util.rsc [hunk 1]
```diff
@@ -37,18 +37,20 @@ bool tryParse(Grammar g, Symbol s, str input, bool allowAmbiguity = false) {
...
 
-bool isRecursive(Grammar g, Symbol s) {
-    set[Symbol] getChildren(Symbol s) 
-        = {s | p <- lookup(g, s), /Symbol s := p.symbols};
+bool isRecursive(Grammar g, Symbol s, set[Symbol] checking = {})
+    = s in checking
+    ? true
+    : any(p <- prodsOf(g, delabel(s)),
+          /Symbol child := p.symbols,
+          isRecursive(g, child, checking = checking + s));
```
Comment (sungshik) [inline rascal-textmate-core/src/main/rascal/lang/rascal/grammar/Util.rsc:46]: This is just a refactoring to make the code simpler (no more nested functions)

File: rascal-textmate-core/src/main/rascal/lang/textmate/conversiontests/NestedCategories.rsc
Comment (sungshik) [inline rascal-textmate-core/src/main/rascal/lang/textmate/conversiontests/NestedCategories.rsc:1]: This new series of tests checks the intended highlights when four symbols with/without categories are organized in a diamond like this:

```
  A
 / \
B   C
 \ /
  D
```

The `.rsc` file defines the grammar; the `.test` file (below) checks the tokenizations

**Comment by DavyLandman** [comment]
> I enabled "Trim Trailing Whitespace", so maybe better to turn off whitespace in the diffs.

please commit an editor config file (and mark it as a suggested extension in the `.vscode` folder

File: rascal-textmate-core/src/main/rascal/lang/textmate/conversiontests/NestedCategories.rsc [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+lexical B11 =               D11 [\ ] "bar11";
+lexical C11 =               D11 [\ ] "baz11";
+lexical D11 = @category="d" "foo11";
+
+lexical A12 =               B12 [\ ] C12;
+lexical B12 =               D12 [\ ] "bar12";
+lexical C12 =               D12 [\ ] "baz12";
+lexical D12 =               "foo12";
+
+Grammar rsc = preprocess(grammar(#Start));
```
Comment (DavyLandman) [inline rascal-textmate-core/src/main/rascal/lang/textmate/conversiontests/NestedCategories.rsc:78]: ouch, what a grammar 🗡️

**Review by DavyLandman — APPROVED** [review]

**Pull request merged**